javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
   def end(): Int = start() + group().length
   def group(): String = ensureLastMatch(0).get
 
+  lazy val startOf: (/*Group number*/Int => /*Start position*/Int) = {
+    // We wrap every char in the regexp in a group to find the intermediate positions.
+    // Same for subgroups.
+    val regex = pattern0.toString
+    // (?:AB)C => (?:(AB))(?:(C))
+    // Transform:  (A(B))C(D)  to  (((A)(B))(C)(D)) so that using the length of the 1st group and the 3rd, we can obtain the start position of D.
+    
+    /** @param originalGroup Either the 1-based index of the original group that this node encloses, or the modifiers which are present before and after (ex: ?, *?, (?=...), (?!..))
+      * @param 
+      */
+    case class Node(originalGroup: Either[Int, (String, String)], regex: Either[String, Seq[Node]]) {
```

Comment:
Those `Either`s seem weird. Why is this not an ADT with various `case class`es, rather than a single case class with `Either` fields?

Reply:
Replaced them by new more explicit classes with comments

```scala
// parentheses aspect of the regex
    sealed trait OriginalRegex {
      @inline def fold[A](numberF: Int => A, leftRightF: (String, String) => A): A = this match {
        case OriginallyGroupped(number) => numberF(number)
        case OriginallyWrapped(left, right) => leftRightF(left, right)
      }
    }
    // If the regex had originally a group number, that is, with simple parentheses
    // non-grouping parentheses (?: ...) are always added
    case class OriginallyGroupped(number: Int) extends OriginalRegex
    // If the regex was not originally groupped with parentheses,
    // then right holds the left and right modifiers (e.g. *, +, *?, +?, ?)
    case class OriginallyWrapped(left: String, right: String) extends OriginalRegex
    
    // Type of the node, a regex or a sequence of nodes.
    sealed trait NodeType {
      @inline def fold[A](regexF: String => A, childrenF: Seq[Node] => A): A = this match {
        case RegexLeaf(regex) => regexF(regex)
        case ParentNode(children) => childrenF(children)
      }
    }
    // A pure regex, without any subgroups
    case class RegexLeaf(regex: String) extends NodeType
    // A parent groupping, with children node. Modifiers to this group, such as parentheses, can be added using the field of Node originalGroup.
    case class ParentNode(children: Seq[Node]) extends NodeType
    
    /** @param originalGroup Either the 1-based index of the original group that this node encloses, or the modifiers which are present before and after (ex: ?, *?, (?=...), (?!..))
      * @param 
      */
    case class Node(originalGroup: OriginalRegex, nodeType: NodeType) {
```